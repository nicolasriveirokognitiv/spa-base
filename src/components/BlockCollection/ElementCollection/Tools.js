import ToolsetController from './ToolsetController';
import Title from './Title';
import Text from './Text';
import Description from './Description';
import BackgroundColor from './BackgroundColor';
import Border from './Border';
import BorderRadius from './BorderRadius';
import TextAlignment from './TextAlignment';
import Width from './Width';
import URL from './URL';
import FlexDirection from './FlexDirection';
import FlexAlignment from './FlexAlignment';
import UploadImage from './UploadImage';

const Tools = {
    Controller    : ToolsetController,
    Title,
    Text,
    Description,
    BackgroundColor,
    Border,
    BorderRadius,
    TextAlignment,
    Width,
    URL,
    FlexDirection,
    FlexAlignment,
    UploadImage,
    defaultValues : {
        title           : Title.defaultProps.defaultValue,
        text            : Text.defaultProps.defaultValue,
        description     : Description.defaultProps.defaultValue,
        backgroundColor : BackgroundColor.defaultProps.defaultValue,
        border          : Border.defaultProps.defaultValue,
        borderRadius    : BorderRadius.defaultProps.defaultValue,
        textAlignment   : TextAlignment.defaultProps.defaultValue,
        width           : Width.defaultProps.defaultValue,
        url             : URL.defaultProps.defaultValue,
        flexDirection   : FlexDirection.defaultProps.defaultValue,
        flexAlignment   : FlexAlignment.defaultProps.defaultValue,
        uploadImage     : UploadImage.defaultProps.defaultValue,
    },
};

// const defaultValues = Object.values(Tools)
//     .filter(tool => tool.defaultProps.defaultValue)
//     .reduce((prev, tool) => ({ ...prev, [tool.defaultProps.propertyKey]: tool.defaultProps.defaultValue }));

// export { defaultValues };

export default Tools;
