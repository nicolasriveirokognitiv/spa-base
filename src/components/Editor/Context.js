import React, { useMemo, useState } from 'react';
import PropTypes from 'prop-types';
import { pageStructurePropTypes } from 'Constants/propTypes';
import EditorContext from 'State/editorContext';
import BlocksContext from 'State/blocksContext';
import EditorInformation from './EditorInformation';

function shouldActivateTestMode() {
    if (process.env.NODE_ENV !== 'production') {
        const searchParams = new URLSearchParams(window.location.search);
        return searchParams.get('qa') !== null;
    }
    return false;
}

function Context({ page, instanciatedBlocks, setInstanciatedBlocks, children }) {
    // The separation between contexts allow us to perform changes in
    // one avoiding to re-render the children of the other one
    const [pageForTesting, setPageForTesting] = useState();

    const editorContext = useMemo(() => ({
        page                 : pageForTesting || page,
        renderEngineMode     : 'static',
        isTestModeActive     : shouldActivateTestMode(),
        isContextInitialized : true,
    }),
    [page, pageForTesting]);

    const blocksContext = useMemo(() => ({
        instanciatedBlocks,
        setInstanciatedBlocks,
        isContextInitialized: true,
    }),
    [instanciatedBlocks, setInstanciatedBlocks]);

    return (
        <BlocksContext.Provider value={blocksContext}>
            <EditorContext.Provider value={editorContext}>
                {editorContext.isTestModeActive && (
                    <EditorInformation
                        setPageForTesting={setPageForTesting}
                        info={{ title: editorContext.page.config.title, request: {}, unsavedBlocks: '', instanciatedBlocks }}
                        loadedPage={page}
                    />
                )}
                {children}
            </EditorContext.Provider>
        </BlocksContext.Provider>
    );
}

Context.propTypes = {
    page                  : pageStructurePropTypes.isRequired,
    instanciatedBlocks    : PropTypes.shape({}).isRequired,
    setInstanciatedBlocks : PropTypes.func.isRequired,
    children              : PropTypes.node.isRequired,
};

export default Context;
