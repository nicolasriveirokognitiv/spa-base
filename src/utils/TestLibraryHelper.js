/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import initStore from 'State/store';

const store = initStore();

const renderWithRedux = children => render(
    <Provider store={store}>
        {children}
    </Provider>,
);

export * from '@testing-library/react';
export { renderWithRedux, render };
