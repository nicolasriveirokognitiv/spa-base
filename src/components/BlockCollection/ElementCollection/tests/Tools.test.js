import React from 'react';
import Tools from '../Tools';
import Context from 'Components/Editor/Context';
import { render, fireEvent } from 'Utils/TestLibraryHelper';

const blockId = 'a';
const blockType = 'card';
const instanciatedBlocksMock = {
    [blockId]: {
        current: {
            setProperties     : jest.fn(),
            properties        : { data: { }, config: {}, design: {} },
            unsavedProperties : { },
        },
    },
};
const littlePageMock = {
    config: {
        title           : '',
        metaDescription : '',
        metaKeywords    : '',
    },
    design : { className: '' },
    rows   : {
        columns: [{
            blocks: [
                {
                    type     : 'Card',
                    category : 'Basic',
                    config   : {},
                    data     : {},
                    design   : {},
                    content  : {},
                },
            ],
        }],
    },
};

// Enables QA Mode on the Context
window.history.pushState({}, 'Test QA', '?qa');

describe('Tools', () => {
    let props;
    let container;

    beforeEach(() => {
        container = document.createElement('div');
        document.body.appendChild(container);
        props = { blockId, blockType };
    });

    const renderController = (setBlocks = () => {}) => render(
        <Context
            instanciatedBlocks={instanciatedBlocksMock}
            setInstanciatedBlocks={setBlocks}
            page={littlePageMock}
        >
            <Tools.Controller {...props}>
                <Tools.Title />
            </Tools.Controller>
        </Context>,
    );

    test('Should render the controller correctly', () => {
        expect(renderController()).toBeDefined();
    });

    test('Should match snapshot for the controller', async () => {
        const { asFragment } = renderController();
        expect(asFragment()).toMatchSnapshot();
    });

    test('Should call setProperties', async () => {
        const { getByTestId } = renderController();
        const element = getByTestId('card-data-title');

        fireEvent.click(element);
        fireEvent.change(element, { target: { value: 'foo' } });

        expect(instanciatedBlocksMock[blockId].current.setProperties).toHaveBeenCalled();
    });
});
