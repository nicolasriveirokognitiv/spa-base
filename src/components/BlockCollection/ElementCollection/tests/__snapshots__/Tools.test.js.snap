// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tools Should match snapshot for the controller 1`] = `
<DocumentFragment>
  <div
    class="qa-button-wrapper"
    data-instanciated-blocks-ids="a"
    data-loaded-page-json="{\\"config\\":{\\"title\\":\\"\\",\\"metaDescription\\":\\"\\",\\"metaKeywords\\":\\"\\"},\\"design\\":{\\"className\\":\\"\\"},\\"rows\\":{\\"columns\\":[{\\"blocks\\":[{\\"type\\":\\"Card\\",\\"category\\":\\"Basic\\",\\"config\\":{},\\"data\\":{},\\"design\\":{},\\"content\\":{}}]}]}}"
    data-loaded-page-title=""
    data-testid="editor-information"
    data-unsaved-blocks=""
  >
    <div
      class="qa-button-data"
    >
      <div
        class="qa-button-data-title"
      >
        Editor Information
      </div>
      <div
        class="qa-button-data-block"
      >
        <span
          class="qa-button-data-label"
        >
          Logged User:
        </span>
        <span
          class="qa-button-data-content"
        >
          -
        </span>
      </div>
      <div
        class="qa-button-data-block"
      >
        <span
          class="qa-button-data-label"
        >
          Request state:
        </span>
        <span
          class="qa-button-data-content"
        >
          -
        </span>
      </div>
      <div
        class="qa-button-data-block"
      >
        <span
          class="qa-button-data-label"
        >
          Request URL:
        </span>
        <span
          class="qa-button-data-content"
        >
          -
        </span>
      </div>
      <div
        class="qa-button-data-block"
      >
        <span
          class="qa-button-data-label"
        >
          Request Response:
        </span>
        <span
          class="qa-button-data-content"
        >
          -
        </span>
      </div>
      <div
        class="qa-button-data-block"
      >
        <span
          class="qa-button-data-label"
        >
          Unsaved Blocks:
        </span>
        <span
          class="qa-button-data-content"
        >
          -
        </span>
      </div>
      <div
        class="qa-button-data-block"
      >
        <span
          class="qa-button-data-label"
        >
          Instanciated Blocks  ID's:
        </span>
        <span
          class="qa-button-data-content"
        >
          a
        </span>
      </div>
      <div
        class="qa-button-data-title"
      >
        Loaded Page
      </div>
      <div
        class="qa-button-data-block"
      >
        <span
          class="qa-button-data-label"
        >
          Page title:
        </span>
        <span
          class="qa-button-data-content"
        />
      </div>
      <div
        class="qa-button-data-block"
      >
        <span
          class="qa-button-data-label"
        >
          External JSON:
        </span>
        <textarea
          class="ant-input qa-button-data-content"
          data-testid="editor-information-loaded-json"
          readonly=""
          rows="3"
        >
          {"config":{"title":"","metaDescription":"","metaKeywords":""},"design":{"className":""},"rows":{"columns":[{"blocks":[{"type":"Card","category":"Basic","config":{},"data":{},"design":{},"content":{}}]}]}}
        </textarea>
      </div>
      <div
        class="qa-button-data-block"
      >
        <span
          class="qa-button-data-label"
        >
          Internal JSON Mock:
        </span>
        <textarea
          class="ant-input"
          data-testid="editor-information-json-mock"
          placeholder="Paste a new JSON Page Structure here to replace the loaded one"
          rows="6"
        />
      </div>
      <div
        class="qa-button-data-block text-align-right"
      >
        <button
          class="ant-btn qa-button-data-button ant-btn-text"
          data-testid="editor-information-refresh"
          type="button"
        >
          <span>
            Clean mock
          </span>
        </button>
        <button
          class="ant-btn qa-button-data-button ant-btn-text"
          data-testid="editor-information-refresh"
          type="button"
        >
          <span>
            Refresh
          </span>
        </button>
      </div>
    </div>
  </div>
  <div
    class="page-control-item-tabs"
  >
    <div
      class="ant-tabs ant-tabs-top ant-tabs-line"
    >
      <div
        class="ant-tabs-bar ant-tabs-top-bar"
        role="tablist"
        tabindex="-1"
      >
        <div
          class="ant-tabs-nav-container"
        >
          <span
            class="ant-tabs-tab-prev ant-tabs-tab-btn-disabled"
            unselectable="unselectable"
          >
            <span
              class="ant-tabs-tab-prev-icon"
            >
              <span
                aria-label="left"
                class="anticon anticon-left ant-tabs-tab-prev-icon-target"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  class=""
                  data-icon="left"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                  />
                </svg>
              </span>
            </span>
          </span>
          <span
            class="ant-tabs-tab-next ant-tabs-tab-btn-disabled"
            unselectable="unselectable"
          >
            <span
              class="ant-tabs-tab-next-icon"
            >
              <span
                aria-label="right"
                class="anticon anticon-right ant-tabs-tab-next-icon-target"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  class=""
                  data-icon="right"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                  />
                </svg>
              </span>
            </span>
          </span>
          <div
            class="ant-tabs-nav-wrap"
          >
            <div
              class="ant-tabs-nav-scroll"
            >
              <div
                class="ant-tabs-nav ant-tabs-nav-animated"
              >
                <div>
                  <div
                    aria-controls="tabpane-1"
                    aria-disabled="false"
                    aria-selected="true"
                    class="ant-tabs-tab-active ant-tabs-tab"
                    id="tab-1"
                    role="tab"
                    tabindex="0"
                  >
                    Content
                  </div>
                  <div
                    aria-controls="tabpane-2"
                    aria-disabled="false"
                    aria-selected="false"
                    class=" ant-tabs-tab"
                    id="tab-2"
                    role="tab"
                    tabindex="-1"
                  >
                    Style
                  </div>
                  <div
                    aria-controls="tabpane-3"
                    aria-disabled="false"
                    aria-selected="false"
                    class=" ant-tabs-tab"
                    id="tab-3"
                    role="tab"
                    tabindex="-1"
                  >
                    Settings
                  </div>
                </div>
                <div
                  class="ant-tabs-ink-bar ant-tabs-ink-bar-animated"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-tabs-content ant-tabs-content-animated ant-tabs-top-content"
        style="margin-left: 0%;"
      >
        <div
          aria-hidden="false"
          class="ant-tabs-tabpane ant-tabs-tabpane-active"
          role="tabpanel"
          style="visibility: visible;"
          tabindex="0"
        >
          <div
            class="tool-wrapper"
          >
            <label
              class="tool-input-label"
              data-testid="label-title"
            >
              Title
            </label>
            <input
              class="ant-input"
              data-testid="card-data-title"
              placeholder="Title"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          aria-hidden="true"
          class="ant-tabs-tabpane ant-tabs-tabpane-inactive"
          role="tabpanel"
          style="visibility: hidden;"
          tabindex="-1"
        />
        <div
          aria-hidden="true"
          class="ant-tabs-tabpane ant-tabs-tabpane-inactive"
          role="tabpanel"
          style="visibility: hidden;"
          tabindex="-1"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
